src/types.rs

```diff
@@ -488,6 +500,410 @@ impl ZipFileData {
+            size,
+            uncompressed_size,
+            compressed_size,
+            header_start,
+        })
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(packed)]
+pub(crate) struct ZipCentralEntryBlock {
+    pub magic: spec::Magic,
+    pub version_made_by: u16,
+    pub version_to_extract: u16,
+    pub flags: u16,
+    pub compression_method: u16,
+    pub last_mod_time: u16,
+    pub last_mod_date: u16,
+    pub crc32: u32,
+    pub compressed_size: u32,
+    pub uncompressed_size: u32,
+    pub file_name_length: u16,
+    pub extra_field_length: u16,
+    pub file_comment_length: u16,
+    pub disk_number: u16,
+    pub internal_file_attributes: u16,
+    pub external_file_attributes: u32,
+    pub offset: u32,
+}
+
+impl Block for ZipCentralEntryBlock {
+    const MAGIC: spec::Magic = spec::Magic::CENTRAL_DIRECTORY_HEADER_SIGNATURE;
+
+    #[inline(always)]
+    fn magic(self) -> spec::Magic {
+        self.magic
+    }
+
+    const ERROR: ZipError = ZipError::InvalidArchive("Invalid Central Directory header");
+
+    to_and_from_le![
+        (magic, spec::Magic),
+        (version_made_by, u16),
+        (version_to_extract, u16),
+        (flags, u16),
+        (compression_method, u16),
+        (last_mod_time, u16),
+        (last_mod_date, u16),
+        (crc32, u32),
+        (compressed_size, u32),
+        (uncompressed_size, u32),
+        (file_name_length, u16),
+        (extra_field_length, u16),
+        (file_comment_length, u16),
+        (disk_number, u16),
+        (internal_file_attributes, u16),
+        (external_file_attributes, u32),
+        (offset, u32),
+    ];
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(packed)]
+pub(crate) struct ZipLocalEntryBlock {
+    magic: spec::Magic,
+    pub version_made_by: u16,
+    pub flags: u16,
+    pub compression_method: u16,
+    pub last_mod_time: u16,
+    pub last_mod_date: u16,
+    pub crc32: u32,
+    pub compressed_size: u32,
+    pub uncompressed_size: u32,
+    pub file_name_length: u16,
+    pub extra_field_length: u16,
+}
+
+impl Block for ZipLocalEntryBlock {
+    const MAGIC: spec::Magic = spec::Magic::LOCAL_FILE_HEADER_SIGNATURE;
+
+    #[inline(always)]
+    fn magic(self) -> spec::Magic {
+        self.magic
+    }
+
+    const ERROR: ZipError = ZipError::InvalidArchive("Invalid local file header");
+
+    to_and_from_le![
+        (magic, spec::Magic),
+        (version_made_by, u16),
+        (flags, u16),
+        (compression_method, u16),
+        (last_mod_time, u16),
+        (last_mod_date, u16),
+        (crc32, u32),
+        (compressed_size, u32),
+        (uncompressed_size, u32),
+        (file_name_length, u16),
+        (extra_field_length, u16),
+    ];
+}
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct Zip64ExtraFieldBlock {
+    magic: spec::ExtraFieldMagic,
+    size: u16,
+    uncompressed_size: Option<u64>,
+    compressed_size: Option<u64>,
+    header_start: Option<u64>,
```

Comment:
I decided not to try to fix the `parse_extra_field()` code right now, especially since it deals with AES data, which I've decided not to touch in this PR. I've left some TODOs about this: https://github.com/zip-rs/zip2/blob/2cf03934f414a86cf521dae33129936d4d199ce8/src/spec.rs#L64 https://github.com/zip-rs/zip2/blob/d6ea5ed6fe152891a9ab7f9941fe7fcb283a7e6b/src/read.rs#L1093-L1115 

I think your proposed change would make it more difficult to follow up and improve the `parse_extra_field()` code in a subsequent PR. I also think retaining the individual fields serves the purpose of this PR better, which is to:
1. unify parsing/serialization into a single struct,
2. decouple setting fields from serialization,
3. replace a bunch of imperative `.read_*_le()` calls with calls to clearly-named methods.

With that in mind, I would like to keep the individual fields here, even if we can't merge them into the same `Block` trait since the size isn't constant. What are your thoughts on that?

Reply:
Have reverted this in the most recent commit.